os/kernel/binary_manager/binary_manager_binfile.c

```diff
@@ -0,0 +1,215 @@
+/****************************************************************************
+ *
+ * Copyright 2020 Samsung Electronics All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND,
+ * either express or implied. See the License for the specific
+ * language governing permissions and limitations under the License.
+ *
+ ****************************************************************************/
+/****************************************************************************
+ * Included Files
+ ****************************************************************************/
+#include <tinyara/config.h>
+#include <debug.h>
+#include <dirent.h>
+#include <stdio.h>
+#include <string.h>
+#include <unistd.h>
+#include <fcntl.h>
+#include <errno.h>
+#include <stdlib.h>
+#include <limits.h>
+#include <sys/stat.h>
+#include <tinyara/binary_manager.h>
+
+#include "binary_manager/binary_manager.h"
+
+/****************************************************************************
+ * Private Functions
+ ****************************************************************************/
+/****************************************************************************
+ * Name: binary_manager_clear_binfile
+ *
+ * Description:
+ *	 This function clears binary files with old version.
+ *
+ ****************************************************************************/
+static int binary_manager_clear_binfile(int bin_idx)
```

Comment:
you already defined naming rule of binary.
In this case you'd better load binaries when this initialized, and manage it with some structure.
Then you don't have to use opendir ~ readdir any more...
```
//when init
void scan_binary() {
    dirp = opendir();
    if (!dirp) {
            mkdir(); // to avoid your patterns( open(O_CREAT) -> failed but -ENOENT -> mkdir -> open(O_CREAT)...)
     } else {
            for (;;) {
                  make_binary_map_here(); //but I think you already have a similar one.
             }
             sort_by_version();
}

void binary_manager_load_binary() {
  ....
  set_running_binary_map(); //set running binary in binary map
}
```
when new binary comes, you can remove specific one and add new binary, and then sort.
It is very easy because binary map will show name and running state.
Also if bin_idx is not necessary actually if is just sequence number, not version